cmd/block-generator/generator/generate.go

```diff
@@ -290,35 +371,193 @@ func (g *generator) generatePaymentTxn(round uint64, intra uint64) (types.Signed
-			Note:        nil,
-			GenesisID:   g.genesisID,
-			GenesisHash: g.genesisHash,
-			Group:       sdk_types.Digest{},
-			Lease:       [32]byte{},
-			RekeyTo:     sdk_types.Address{},
-		},
-		KeyregTxnFields: sdk_types.KeyregTxnFields{},
-		PaymentTxnFields: sdk_types.PaymentTxnFields{
-			Receiver:         receiver,
-			Amount:           sdk_types.MicroAlgos(amount),
-			CloseRemainderTo: sdk_types.Address{},
-		},
-		AssetConfigTxnFields:   sdk_types.AssetConfigTxnFields{},
-		AssetTransferTxnFields: sdk_types.AssetTransferTxnFields{},
-		AssetFreezeTxnFields:   sdk_types.AssetFreezeTxnFields{},
-		ApplicationFields:      sdk_types.ApplicationFields{},
-	}
 
+	// TODO: Unnecessary serialize/deserialize of addresses.
+	// Easy to optimize by creating the struct ourselves if needed.
+	txn, err := future.MakePaymentTxn(sender.String(), receiver.String(), amount, nil, "", sp)
+	util.MaybeFail(err, "unable to make transaction")
 	return convert(txn, types.ApplyData{}), nil
 }
 
-func (g *generator) generateAssetTxn(round uint64, intra uint64) (types.SignedTxnWithAD, error) {
-	return convert(sdk_types.Transaction{}, types.ApplyData{}), nil
+func getAssetTxOptions() []interface{} {
+	return []interface{}{assetCreate, assetDestroy, assetOptin, assetXfer, assetClose}
+}
+
+func (g *generator) generateAssetTxnInternal(txType interface{}, sp sdk_types.SuggestedParams, intra uint64) (txn types.Transaction) {
+	return g.generateAssetTxnInternalHint(txType, sp, intra, 0, nil)
+}
+
+func (g *generator) generateAssetTxnInternalHint(txType interface{}, sp sdk_types.SuggestedParams, intra uint64, hintIndex uint64, hint *assetData) (txn types.Transaction) {
+	// If there are no assets the next operation needs to be a create.
+	if len(g.assets) == 0 {
+		txType = assetCreate
+	}
+
+	var err error
+	numAssets := uint64(len(g.assets))
+	var senderIndex uint64
+
+	if txType == assetCreate {
+		senderIndex = numAssets % g.numAccounts
+		senderAcct := indexToAccount(senderIndex)
+		senderAcctStr := senderAcct.String()
+
+		total := assetTotal
+		assetID := g.txnCounter + intra + 1
+		assetName := fmt.Sprintf("asset #%d", assetID)
+		txn, err = future.MakeAssetCreateTxn(senderAcctStr, nil, sp, total, 0, false, senderAcctStr, senderAcctStr, senderAcctStr, senderAcctStr, "tokens", assetName, "https://algorand.com", "metadata!")
+		util.MaybeFail(err, "unable to make asset create transaction")
+
+		// Compute asset ID and initialize holdings
+		a := assetData{
+			assetID:  assetID,
+			creator:  senderIndex,
+			holdings: []assetHolding{{acctIndex: senderIndex, balance: total}},
+			holders:  map[uint64]bool{senderIndex: true},
+		}
+
+		g.assets = append(g.assets, &a)
+	} else {
+		assetIndex := rand.Uint64() % numAssets
+		asset := g.assets[assetIndex]
+		if hint != nil {
+			assetIndex = hintIndex
+			asset = hint
+		}
+
+		switch txType {
+		case assetDestroy:
+			// delete asset
+
+			// If the creator doesn't have all of them, close instead
```

Comment:
How likely is it that an asset is actually destroyed? I'm worried that most of the time we will close it instead.

Reply:
Yeah this is a big downside to the approach of having general distributions. If the destroy test also has asset transfers then it wont really test what we want.

My thinking was that this test would need to have a distribution where 50% of the time it created an asset and 50% of the time it deletes the asset.

Maybe this case should be an error instead. Or maybe there should be a report at the end so that we know the actual transaction distributions match what we wanted to test.

Reply:
Adding a report endpoint in another PR